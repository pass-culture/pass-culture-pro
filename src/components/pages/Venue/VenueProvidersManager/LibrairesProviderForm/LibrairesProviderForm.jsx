import React, {PureComponent} from 'react'
import {Form} from 'react-final-form'
import PropTypes from 'prop-types'

class LibrairesProviderForm extends PureComponent {
  constructor() {
    super()
    this.state = {
      isLoadingMode: false,
    }
  }

  handleFormSubmit = () => {
    this.setState({ isLoadingMode: true })

    const { createVenueProvider } = this.props
    return createVenueProvider(this.handleFail, this.handleSuccess)
  }

  handleSuccess = () => {
    const {
      history,
      offererId,
      venueId,
    } = this.props
    history.push(`/structures/${offererId}/lieux/${venueId}`)
  }

  handleFail = (state, action) => {
    const { notify } = this.props
    const {
      payload: { errors },
    } = action

    notify(errors)
  }

  renderForm = props => {
    const { venueSiret } = this.props
    const { isLoadingMode } = this.state

    return (
      <form onSubmit={props.handleSubmit}>
        <div className="libraires-provider-form">
          <div className="compte-section">
            <div className="account-label">
              {'Compte'}
            </div>
            <div
              className='account-value'
            >
              {venueSiret}
            </div>
          </div>

          {!isLoadingMode && (
            <div className="provider-import-button-container">
              <button
                className="button is-intermediate provider-import-button"
                type="submit"
              >
                {'Importer'}
              </button>
            </div>
          )}
        </div>
      </form>
    )
  }

  render() {
    return (
      <Form
        onSubmit={this.handleFormSubmit}
        render={this.renderForm}
      />)
  }
}

LibrairesProviderForm.propTypes = {
  createVenueProvider: PropTypes.func.isRequired,
  history: PropTypes.shape().isRequired,
  notify: PropTypes.func.isRequired,
  providerId: PropTypes.string.isRequired,
  venueId: PropTypes.string.isRequired,
  venueSiret: PropTypes.string.isRequired
}

export default LibrairesProviderForm
