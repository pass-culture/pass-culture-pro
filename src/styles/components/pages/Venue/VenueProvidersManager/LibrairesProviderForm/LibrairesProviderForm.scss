.libraires-provider-form {
  display: flex;
  flex-direction: row;

  .compte-section {
    flex-direction: column;
    max-width: 250px;
    margin-right: 25px;

    .account-label {
      margin-top: 6px;
      margin-bottom: 15px;
    }

    .account-value {
      font-size: 1.2rem;
      font-weight: 500;
      height: 36px;
    }
  }

  .provider-import-button-container {
    position: relative;

    button {
      position: absolute;
      top: 50%;
    }
  }
}
