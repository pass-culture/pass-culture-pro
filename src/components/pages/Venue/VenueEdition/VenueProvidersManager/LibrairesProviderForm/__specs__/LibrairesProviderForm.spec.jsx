import React from 'react'
import { mount, shallow } from 'enzyme'

import LibrairesProviderForm from '../LibrairesProviderForm'
import { Form } from 'react-final-form'

describe('src | components | pages | Venue | VenueProvidersManager | form | LibrairesProviderForm', () => {
  let cancelProviderSelection
  let createVenueProvider
  let props
  let notify
  let history

  beforeEach(() => {
    cancelProviderSelection = jest.fn()
    createVenueProvider = jest.fn()
    history = {
      push: jest.fn(),
    }
    notify = jest.fn()
    props = {
      cancelProviderSelection,
      createVenueProvider,
      history,
      notify,
      offererId: 'CC',
      providerId: 'CC',
      venueId: 'AA',
      venueIdAtOfferProviderIsRequired: false,
      venueSiret: '12345678912345',
    }
  })

  it('should match the snapshot', () => {
    // when
    const wrapper = shallow(<LibrairesProviderForm {...props} />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  it('should initialize LibrairesProviderForm component with default state', () => {
    // when
    const wrapper = shallow(<LibrairesProviderForm {...props} />)

    // then
    expect(wrapper.state()).toStrictEqual({
      isLoadingMode: false,
    })
  })

  describe('when not in loading mode', () => {
    it('should display an import button', () => {
      // when
      const wrapper = mount(<LibrairesProviderForm {...props} />)

      // then
      const importButtonContainer = wrapper.find('.provider-import-button-container')
      expect(importButtonContainer).toHaveLength(1)
      const importButton = importButtonContainer.find('button')
      expect(importButton).toHaveLength(1)
      expect(importButton.prop('className')).toBe('button is-intermediate provider-import-button')
      expect(importButton.prop('type')).toBe('submit')
      expect(importButton.text()).toBe('Importer')
    })

    it('should render the title of the section compte', () => {
      // when
      const wrapper = mount(<LibrairesProviderForm {...props} />)

      // then
      const form = wrapper.find(Form)
      expect(form).toHaveLength(1)
      const label = form.find('.account-label')
      expect(label.text()).toBe('Compte')
    })

    it('should display the venue siret as provider identifier', () => {
      // when
      const wrapper = mount(<LibrairesProviderForm {...props} />)

      // then
      const form = wrapper.find(Form)
      expect(form).toHaveLength(1)
      const textField = form.find('.account-value')
      expect(textField).toHaveLength(1)
      expect(textField.text()).toBe('12345678912345')
    })
  })

  describe('handleSubmit', () => {
    it('should update venue provider using API', () => {
      // given
      const formValues = {
        preventDefault: jest.fn(),
      }
      const wrapper = shallow(<LibrairesProviderForm {...props} />)

      // when
      wrapper.instance().handleFormSubmit(formValues, {})

      // then
      expect(wrapper.state('isLoadingMode')).toBe(true)
      expect(createVenueProvider).toHaveBeenCalledWith(expect.any(Function), expect.any(Function))
    })
  })

  describe('handleSuccess', () => {
    it('should update current url when action was handled successfully', () => {
      // given
      const wrapper = shallow(<LibrairesProviderForm {...props} />)

      // when
      wrapper.instance().handleSuccess()

      // then
      expect(history.push).toHaveBeenCalledWith('/structures/CC/lieux/AA')
    })
  })

  describe('handleFail', () => {
    it('should display a notification with the proper informations', () => {
      // given
      const wrapper = shallow(<LibrairesProviderForm {...props} />)
      const action = {
        payload: {
          errors: [
            {
              error: 'fake error',
            },
          ],
        },
      }

      // when
      wrapper.instance().handleFail({}, action)

      // then
      expect(notify).toHaveBeenCalledWith([{ error: 'fake error' }])
      expect(cancelProviderSelection).toHaveBeenCalledTimes(1)
    })
  })
})
