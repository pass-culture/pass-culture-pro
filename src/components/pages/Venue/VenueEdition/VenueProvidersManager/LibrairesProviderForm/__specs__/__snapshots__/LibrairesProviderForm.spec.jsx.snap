// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | Venue | VenueProvidersManager | form | LibrairesProviderForm should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LibrairesProviderForm
    cancelProviderSelection={[MockFunction]}
    createVenueProvider={[MockFunction]}
    history={
      Object {
        "push": [MockFunction],
      }
    }
    notify={[MockFunction]}
    offererId="CC"
    providerId="CC"
    venueId="AA"
    venueIdAtOfferProviderIsRequired={false}
    venueSiret="12345678912345"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "onSubmit": [Function],
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "onSubmit": [Function],
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
